pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go

```diff
@@ -72,7 +79,70 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
 	return nil
 }
 
+// AddCount increases 1 count for waitGroup
+func (c FakeClient) AddCount() {
+	c.waitGroup.Add(1)
+}
+
+// DecreaseCount decreases 1 count for waitGroup
+func (c FakeClient) DecreaseCount() {
+	c.waitGroup.Done()
+}
+
 // WithClient adds to the context a fake client with the desired behaviour
 func WithClient(ctx context.Context, behaviour *FakeClientBehaviour) context.Context {
 	return context.WithValue(ctx, ceKey{}, newFakeClient(behaviour))
 }
+
+// CheckCloudEventsUnordered checks that all events in wantEvents, and no others,
+// were received via the given chan, in any order.
+func (c *FakeClient) CheckCloudEventsUnordered(t *testing.T, eventChan chan string, testName string, wantEvents []string) error {
+	t.Helper()
+	c.waitGroup.Wait()
+	err := eventsFromChannelUnordered(eventChan, wantEvents)
+	if err != nil {
+		return fmt.Errorf("error in test %s: %v", testName, err)
+	}
+	return nil
+}
+
+// eventsFromChannelUnordered takes a chan of string and a list of events that a test
+// expects to receive. The events can be received in any order. Any extra or too few
+// events are both considered errors.
+func eventsFromChannelUnordered(c chan string, wantEvents []string) error {
+	timer := time.NewTimer(10 * time.Millisecond)
```

Comment:
I think we can not remove it, it is not related to the blocking, but a timer used to read from channel

Reply:
Why do the tests hang if the timer is not used? The only reason this should happen is if the expected number of events is less than the actual number of events received.

Reply:
It's because the previous code will loop +1 times, and during the extra loop since we don't have any events from channel, and also we don't have a timer, it will just hand there,
```
loop len(expected) + 1 times to catch extra erroneous events received that the test is not expecting
```

Reply:
To fix it we can just reduce the loop to len(expected) not len(expected)+1. But it would be a change to this function's behaviour. I don't know why or who implement this logic. I suggest we if we want to do it, we should leave it to another clean up PR to avoid extra changes in this PR

Reply:
Oh I see-- the reason is because we don't want to make sure extra events have been sent, so we wait a little extra to see if any more events are received. What we really need is for the fake client to be able to indicate that it's done sending events (e.g. by closing the channel), but the fake client doesn't know when events are done being sent because they can only be sent one at a time so I'm not sure if this is possible. Let me think about if there's a better way to do this. Keeping this logic here means that our tests are still not deterministic.

Reply:
Sure! Thanks!

Reply:
ok! I have an idea! We should update [newFakeClient](https://github.com/tektoncd/pipeline/blob/14eba83bf31277dc3f8c91fa798b38e30d0a8e6a/pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go#L43-L48) to initialize the fake client buffer size with len(wantEvents). 

The reason is that sends block on a buffered channel when it is full. We are currently waiting (via waitGroup.Wait()) until all events are sent before reading any events from the channel. This means if the controller sends more events than expected, these sends will block because the channel is full, and Wait() will wait indefinitely.

With this solution, if the client sends more or fewer events than expected, Wait() will block indefinitely, but if the client sends the correct number of events, Wait() will wait until they are all sent, and no longer! a.k.a. no more flakes!

Does this make sense? WDYT? If you implement this please remove the timer and add some commentary on this strategy.

Reply:
Maybe? Let me test this solution. Thank you for coming up with this idea!!

Reply:
I tried to implement this in the latest commit, plz lmk if that is aligned what you're suggesting.